src/crypto/tls/handshake_client_tls13.go

```diff
@@ -34,6 +35,41 @@ type clientHandshakeStateTLS13 struct {
 	trafficSecret []byte // client_application_traffic_secret_0
 }
 
+// processDelegatedCredentialFromServer unmarshals the DelegatedCredential
+// offered by the server (if present) and validates it using the peer
+// certificate.
+func (hs *clientHandshakeStateTLS13) processDelegatedCredentialFromServer(dc []byte) error {
+	c := hs.c
+
+	var dCred *DelegatedCredential
+	var err error
+	if dc != nil {
+		// Assert that the DC extension was indicated by the client.
+		if !hs.hello.delegatedCredentialSupported {
+			c.sendAlert(alertUnexpectedMessage)
+			// TODO: check this error is consistent
+			return errors.New("tls: got delegated credential extension without indication")
+		}
+
+		dCred, err = UnmarshalDelegatedCredential(dc)
+		if err != nil {
+			c.sendAlert(alertDecodeError)
+			return fmt.Errorf("tls: delegated credential: %s", err)
+		}
+	}
+
+	if dCred != nil && !c.config.InsecureSkipVerify {
```

Comment:
I don't think that `InsecureSkipVerify` should be skipped here. That setting will skip verifying the certificate chain, but I think that a server providing a delegated credential that is actually invalid for the provided leaf cert is a bug that should be caught.

Reply:
+1 -- this is a good point. That parameter behaves as follows:

```
    // InsecureSkipVerify controls whether a client verifies the server's
    // certificate chain and host name. If InsecureSkipVerify is true, crypto/tls
    // accepts any certificate presented by the server and any host name in that
    // certificate. In this mode, TLS is susceptible to machine-in-the-middle
    // attacks unless custom verification is used. This should be used only for
    // testing or in combination with VerifyConnection or VerifyPeerCertificate.
    InsecureSkipVerify bool
```

So it seems we still want to verify the handshake signature. (I think it's fine to skip the DC signature here, since the DC is just an extension of the certificate chain, which `InsecureSkipVerify` controls.)

Reply:
Mmm... well what this is specifying is: if `c.config.InsecureSkipVerify` is equal to false (meaning `InsecureSkipVerify` will not be used), then the delegated credential can be validated. It is not really skipped but rather checking that the delegated credential is validate when `c.config.InsecureSkipVerify` is set to false.

Reply:
The user can't set `verifiedDC` from VerifyPeerCertificate, right?  So when talking to a server that uses DCs and setting InsecureSkipVerify to true will cause a mismatch between public key used to check the signature (the leaf) and the one that should be used (the DC).

Reply:
I'm not really sure I follow. The delegated credential will still be used, it is only not validated:

```
	if dCred != nil && !c.config.InsecureSkipVerify {
		if !dCred.Validate(c.peerCertificates[0], "server", c.config.time(), certVerifyMsg) {
			c.sendAlert(alertIllegalParameter)
			return errors.New("tls: invalid delegated credential")
		}
	}

	c.verifiedDC = dCred

```

Reply:
Ah, yes, you're right.  I was mistaken about the signature mismatch.

The issue remains that when we set `InsecureSkipVerify` to true, that a user cannot implement a secure verification themselves as they don't have access to the delegated credential, right?  We'd want an analogue of `VerifyPeerCertificate`, say `VerifyDelegatedCredential`.

Secondly, there might be users that already set `InsecureSkipVerify` to true and implement their own secure verification.  At the moment we would break their implementation because now we could use delegated credentials which they currently do not check.

I'm not sure, but I think it might be best to have a separate `InsecureSkipDCVerify`.

Reply:
Oh, I like that. It can actually be inside the `VerifyPeerCertificate` if a delegated credential is present on it; but as that cannot be enforced, I think it is better to create a `VerifyDelegatedCredential` func.

Reply:
I also think this can be addressed on the draft itself.. what you think @chris-wood ?

Reply:
> Oh, I like that. It can actually be inside the `VerifyPeerCertificate` if a delegated credential is present on it;

Wouldn't changing `VerifyPeerCertificate` break the API?

Reply:
Don't think so. It will be just a check to see `if dCred != nil validate , else keep going`. How will it break compatibility?

Reply:
You would need to add an argument to `VerifyPeerCertificate`, right?

Reply:
Sure. But this stills depends on the `InsecureSkipVerify` and what it does. Right now what is happening is that if it used, the DC validation does not happen. But it can also be that `InsecureSkipVerify` is only for the certificate chain.

Reply:
And that DC's are always validated.

Reply:
Yes, that would be better.

Reply:
We could add a separate function to allow the client to validate the delegated credential themselves if need be.

Reply:
Agreed. Spoke around this on the last call, and decided to remove the check ;)